### PR TITLE
Feat/asserts module

### PR DESCRIPTION
**Pull request by Leonard-Pat**
- created an asserts module and added all asserts (apart from no self call)
- Paramterised some asserts
- included some test function in the argent account contract and testing file

### CODE REVIEW

File: contracts/test_argent_account.cairo [hunk 1]
```diff
@@ -48,3 +48,26 @@ fn erc165_supported_interfaces() {
...
+#[available_gas(20000)]
+fn test_assert_only_self() {
+    let retdata = ArgentAccount::assert_only_self_test();
+    assert(retdata == 1, 'return data should be 1');
+}
+
+#[test]
+#[available_gas(20000)]
+fn assert_correct_tx_version_test() {
+    let tx_version = 1;
```
Comment (gaetbout) [inline contracts/test_argent_account.cairo:62]: Is this useful?

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,4 @@ artifacts
 node.json
 .idea
 cairo/
+target/
```
Comment (delaaxe) [inline .gitignore:9]: why is this needed here?

File: contracts/argent_account.cairo [hunk 3]
```diff
@@ -51,4 +52,25 @@ mod ArgentAccount {
...
+        let guardian = guardian::read();
+        asserts::assert_guardian_set(guardian);
+        a
+    }
+
+    fn assert_correct_tx_version_test(tx_version: felt) -> felt {
+        let a = 1;
+        asserts::assert_correct_tx_version(tx_version);
+        1
+    }
```
Comment (delaaxe) [inline contracts/argent_account.cairo:73]: pls let's not include testing code in the main contract, let's move it in the PR's description if not used immediately. same for @gaetbout's `fn get_contract_address_test`

File: contracts/test_argent_account.cairo [hunk 4]
```diff
@@ -48,3 +48,26 @@ fn erc165_supported_interfaces() {
...
+    let retdata = ArgentAccount::assert_correct_tx_version_test(tx_version);
+    assert(retdata == 1, 'return data should be 1');
+}
+
+#[test]
+#[available_gas(20000)]
+#[should_panic(expected = 'argent: guardian required')]
+fn assert_guardian_set_test() {
+    let retdata = ArgentAccount::assert_guardian_set_test();
+    assert(retdata != 1, 'return data should not be 1');
```
Comment (delaaxe) [inline contracts/test_argent_account.cairo:72]: I would put those in a `test_asserts.cairo` file

File: contracts/argent_account.cairo (hunk 3 above)
Comment (gaetbout) [reply contracts/argent_account.cairo:73]: Was just to showcase how to call it.
But yep it should def be removed asap, I should have put that in the chat, my mistake

Comment (delaaxe) [reply contracts/argent_account.cairo:73]: Digging through chats can be cumbersome, better in the PR's description

Comment (Leonard-Pat) [reply contracts/argent_account.cairo:73]: yeah sure will remove it

File: .gitignore (hunk 2 above)
Comment (Leonard-Pat) [reply .gitignore:9]: for some reason running it generated a target dir that it useless so added it to my gitignore

File: contracts/argent_account.cairo [hunk 5]
```diff
@@ -1,6 +1,7 @@
 #[contract]
 mod ArgentAccount {
     use contracts::dummy_syscalls;
+    use contracts::asserts;
```
Comment (gaetbout) [inline contracts/argent_account.cairo:4]: Should be removed if not used

File: contracts/argent_account.cairo [hunk 6]
```diff
@@ -51,4 +52,5 @@ mod ArgentAccount {
     fn get_contract_address_test() -> felt {
         dummy_syscalls::get_contract_address() 
     }
+
```
Comment (delaaxe) [inline contracts/argent_account.cairo:55]: can you also remove this and the corresponding `use`?

Comment (Leonard-Pat) [reply contracts/argent_account.cairo:54]: yeah sure

**Review by delaaxe — APPROVED** [review]

File: contracts/asserts.cairo [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+
+fn assert_initialized(signer: felt) {
+    // read signer from contract storage 
+    // let signer = signer::read();
+    assert(signer != 0, 'argent: account not initialized');
+}
+
+fn assert_correct_tx_version(tx_version: felt) {
+    let correct_tx_version = 1;
```
Comment (juniset) [inline contracts/asserts.cairo:22]: We should put the transaction version and query version into constants.

File: contracts/asserts.cairo [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    assert(self == caller_address, 'argent: only self');
+}
+
+fn assert_non_reentrant(signer: felt) {
+    let caller_address = dummy_syscalls::get_caller_address();
+    assert(caller_address == 0, 'argent: argent: no reentrant call');
+}
+
+
+fn assert_initialized(signer: felt) {
```
Comment (juniset) [inline contracts/asserts.cairo:15]: Given that it doesn't read storage, this assert doesn't really abstract anything so maybe we should move it to account contract. But let's leave it here for now and decide on that later.

File: contracts/test_asserts.cairo [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+#[test]
+#[available_gas(20000)]
+#[should_panic(expected = 'argent: invalid tx version')]
+fn assert_correct_tx_version_invalidtx_test() {
+    // for now valid tx_version == 1 & 2
+    let tx_version = 4;
+    asserts::assert_correct_tx_version(tx_version);
+    let tx_version = 4;
+    asserts::assert_correct_tx_version(tx_version);
```
Comment (gaetbout) [inline contracts/test_asserts.cairo:27]: Any reason behind theses 2 lines being duplicated?

File: contracts/asserts.cairo [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    // read signer from contract storage 
+    // let signer = signer::read();
+    assert(signer != 0, 'argent: account not initialized');
+}
+
+fn assert_correct_tx_version(tx_version: felt) {
+    assert(tx_version == CORRECT_TX_VERSION ^ tx_version == QUERY_VERSION, 'argent: invalid tx version');
+}
+
+fn assert_guardian_set(guardian: felt) {
```
Comment (gaetbout) [inline contracts/asserts.cairo:28]: this could probably be renamed to smthng more generic like assert_not_zero

Comment (delaaxe) [reply contracts/asserts.cairo:28]: assert non zero might as well be inlined as `assert(value != 0, "custom message")` cause you usually want a more specific error than just "was zero"

File: contracts/asserts.cairo [hunk 11]
```diff
@@ -0,0 +1,32 @@
+use contracts::dummy_syscalls;
+
+const CORRECT_TX_VERSION: felt = 1;
+const QUERY_VERSION: felt = 2;
```
[the remark this answers is not in the document]
Comment (Leonard-Pat) [reply contracts/asserts.cairo:4]: For now i get the issue 2**128 is an unsupported feature

File: contracts/test_asserts.cairo (hunk 9 above)
Comment (Leonard-Pat) [reply contracts/test_asserts.cairo:27]: no mb!

File: contracts/asserts.cairo (hunk 11 above)
[the remark this answers is not in the document]
Comment (delaaxe) [reply contracts/asserts.cairo:4]:
```
const QUERY_VERSION = 340282366920938463463374607431768211456 + TRANSACTION_VERSION; // 2**128 + TRANSACTION_VERSION

File: contracts/asserts.cairo (hunk 10 above)
Comment (Leonard-Pat) [reply contracts/asserts.cairo:28]: should i remove assert guardian / assert initliazed and leave them to be used inline instead ?

Comment (delaaxe) [reply contracts/asserts.cairo:28]: I think it's ok for now so we can get this merged, can always factor things later on if needed

**Review by juniset — APPROVED** [review]
